### ISSUE TITLE
Extending ArrayObject with private properties produces false positives

### ISSUE THREAD

**Issue by discordier** [vimeo/psalm#7983]
When defining a class that extends `ArrayObject`, psalm seems a little confused, as it detects private properties to have a different access level than the very same property in the very same class.
```
ERROR: [OverriddenPropertyAccess](https://psalm.dev/159) - 5:20 - Property Foo::$name has different access level than Foo::$name
```
It even reports some type variance for the private property - despite the fact that `ArrayObject` does not declare any properties.
```
ERROR: [NonInvariantPropertyType](https://psalm.dev/265) - 5:20 - Property Foo::$name has type string, not invariant with ArrayObject::$name of type <empty>
```

I fail to see how to tackle this, as I do not know how I could pin point the culprit.

https://psalm.dev/r/8c3b56d382

**Comment by AndrolGenhald** [comment]
For some reason `ArrayObject` has `'name' => 'string'` in the PropertyMap. Would love to know why that's there in the first place before removing it, but it does seem like the property does not actually exist.

**Comment by discordier** [comment]
Where does it come from?
It is not declared in the [stub](https://github.com/vimeo/psalm/blob/19ac15077f6582112bf21ca24d11f4e799f9e5c7/stubs/CoreGenericClasses.phpstub#L133).

I found this:
https://github.com/vimeo/psalm/blob/3a85f49268f8db12d35cc6c991218ddeea72ed02/dictionaries/PropertyMap.php#L32
which accordingly to comment is "stolen" from [phan/phan/src/Phan/Language/Internal/PropertyMap.php](https://github.com/phan/phan/blob/93c1c287d646d08249b4f5539bfe147550b163d7/src/Phan/Language/Internal/PropertyMap.php#L98).
Sadly neither tells where it originates from.
It appears to be wrong non-the-less.

**Comment by orklah** [comment]
It seems to be generated from phpdoc somehow: https://github.com/phan/phan/blob/93c1c287d646d08249b4f5539bfe147550b163d7/src/Phan/Language/Internal/PropertyMap.php#L56

but it seems wrong to me too. @TysonAndre Do you remember the reason for this? It seems that a lot of iterators seems to have a name property but I don't think it's true. Maybe old documentation artifacts?

**Comment by discordier** [comment]
Traced it down to: https://svn.php.net/repository/phpdoc/en/trunk/reference/spl/arrayobject.xml

It has (from line 69 on):
```xml
<!-- }}} -->
  |  
  | <!-- {{{ If the property is documented below (xml:id=arrayobject.props) use this
  | <classsynopsisinfo role="comment">&Properties;</classsynopsisinfo>
  | <fieldsynopsis>
  | <modifier>public</modifier>
  | <type>string</type>
  | <varname linkend="arrayobject.props.name">name</varname>
  | </fieldsynopsis>
  | }}} -->
```
This is commented XML code, which is simply not handled by the bash one liner embedded in the code.

**Comment by orklah** [comment]
wow congrats, this is a pretty old bug found in both tools (possibly in phpstan too?)!

I think there's a lot of name properties we could drop in there

**Comment by discordier** [comment]
New location is here: https://github.com/php/doc-en/blob/master/reference/spl/arrayobject.xml#L101

**Comment by orklah** [comment]
Are you up for a PR to fix this? :)

**Comment by discordier** [comment]
Working on it.
I'll check the other classes aswell, so please allow some time.

**Comment by discordier** [comment]
While working on this, I am a little confused as some classes are denoted with `-` separation and some have the full namespace as `foo\bar`.
Is it relevant or shall I normalize to `\\`?

PS: I wrote a small XML parser script (using `\SimpleXML`), which handles the extraction, I'd put this in `bin` with the PR, is this ok?

**Comment by AndrolGenhald** [comment]
> While working on this, I am a little confused as some classes are denoted with `-` separation

Can you give an example?


> PS: I wrote a small XML parser script (using `\SimpleXML`), which handles the extraction, I'd put this in `bin` with the PR, is this ok?

Absolutely! My goal is to remove extensions from the callmap over time, but it will definitely still be useful for built in types, and with a bit of work it could provide a very useful automatic comparison between the PHP documentation and reflected extension types.

**Comment by orklah** [comment]
the `-` seems weird yeah, it's not like that in phan either. I'd say normalize to `\\`, let's see if it still pass tests.

Yeah, please do for the parser

**Comment by discordier** [comment]
PR submitted as #8000 (nice number :smile:).

**Comment by orklah** [comment]
this was fixed :)
